### DISCUSSION TITLE
@gemu2015: sml channel math is only supported on frequently updated channels

### DISCUSSION THREAD

**Discussion by h-tro** [arendst/Tasmota#15953] [Q&A] [answered]
@gemu2015 
In the closed Issue ## 15950 you wrote:
**sml channel math is only supported on frequently updated channels. a counter is only updated when the counter value changes and only then the channel math is performed.**

Yes, but that works well.
If the counter updates, a new value is calculated and displayed in the "ABB B21 Energy meter" line.
If a 2nd meter is added, it unfortunately no longer works.
No bug?

In my case, the constant value in the line (1, = M 1+**#110** @1, B21 Energy Meter, KWh, Meter_cnt, 3), is the value of the hardware meter display. This value is to be transported by MQTT to Home Assistant and should be displayed there.
Do you have any further hints?

**Comment by gemu2015** [comment]
here is your modified script without sml channel math

\>D
; additional defines
;#define USE_SML_SCRIPT_CMD
;#define USE_SCRIPT_WEB_DISPLAY
;#define USE_SCRIPT_JSON_EXPORT

offset=110
lval=0
dval=0

\>B
->sensor53 r

\>S
; calc 10 second delta
if upsecs%10==0 {
dval=sml[1]-lval
lval=sml[1]
}


; show values in web ui
\>W
B21 Energy{m}%3(sml[1]+offset)% kWh
B21 Current Power{m}%2dval% W
SDM120-1 Energy{m}%3sml[2]% kWh
SDM120-1 Curr Power{m}%2sml[3]% W

; send calculated values to MQTT
\>J
,"ABB":{
"pow":%3(sml[1]+offset)%,
"cpow":%2dval%
}


; hide all meter values from web ui
\>M 2
+1,4,c,1,-50,ABB
+2,22,m,0,2400,Modbus,21,2,0104000C,0104004A

1,1-0:1.8.0*255(@1000,\*,kWh,Energy_Cnt,3
2,010404ffffffff@i1:1,\*,kWh,SDM120-1 Energy_out,3
2,010404ffffffff@i0:-1,\*,W,SDM120-1 Power_act,2
\#

**Reply by h-tro** [reply]
Thanks, the offset value works.

But the calculation of "B21 Current Power" does not work.
I can't find the problem.
See the screenshot !
![CalPower](https://user-images.githubusercontent.com/33256321/178147810-ae40ba71-36d3-4c47-9fc4-731c5d9edde2.JPG)

**Comment by gemu2015** [comment]
i forgot to scale the delta value. since we have kWh which is 3600 Kwattseconds / 10
we need to multiply with 360

dval=sml[1]-lval*360

the longer the deltaperiod the more accurate the result.

if you are satisfied with 1 minute updates
; calc 60 second delta
if upsecs%60==0 {
dval=sml[1]-lval*60
lval=sml[1]
}

**Reply by h-tro** [reply]
you mean,
if you are satisfied with 1 minute updates
; calc 60 second delta
if upsecs%60==0 {
dval=sml[1]-lval***3**60
lval=sml[1]
}

It seems that only works correctly for the 1st minute.
After that, an incorrect value or zero is displayed again.

Both meter AAB and SDM120 are currently connected to the same inverter. B21 Current Power and SDM120-1 Curr Power must be approximately the same value.

**Reply by gemu2015** [reply]
not 360 in case of minute it must be 60 (60 kW minutes = 1 KWh) 
zero is displayed when there is no change in the counter during the last minute

**Reply by h-tro** [reply]
OK!
; calc 60 second delta
if upsecs%**60**==0 {
**dval=sml[1]-lval*60**
lval=sml[1]
}

I can no longer test. The weather is very cloudy.
The inverter / ABB meter is currently not providing enough pulses.

I added the following line again:
1,=d 1 10 @1,B21 Current Power,W,Power_curr,2
The calculation works and seems to be more accurate.
See the screenshot.

I'll test your calculation again tomorrow when more pluses are delivered.

![Cal3](https://user-images.githubusercontent.com/33256321/178152222-27b34a13-b873-42a7-a697-5503f10d35da.JPG)

**Reply by h-tro** [reply]
Now there are about 2 pulses per minute.

But the calculation is wrong. The result should be 128W.
Shouldn't the value be multiplied by 1000 1kW =1000 W?
The calculation is very imprecise.
Shouldn't the current power be calculated from the time between pulses?

See script and screenshot.
>D
; additional defines
;#define USE_SML_SCRIPT_CMD
;#define USE_SCRIPT_WEB_DISPLAY
;#define USE_SCRIPT_JSON_EXPORT

; offset=16.688
offset=0
lval=0
dval=0


>B
->sensor53 r

>S
; calc 60 second delta
if upsecs%60==0{
dval=sml[1]-lval*60
lval=sml[1]
}

; show values in web ui
>W
B21 Energy{m}%3(sml[1]+offset)% kWh
B21 Current Power_lval{m}%4lval% W
B21 Current Power_dval{m}%4dval% W
SDM120-1 Energy{m}%3sml[3]% kWh
SDM120-1 Curr Power{m}%2sml[4]% W

; send calculated values to MQTT
>J
,"ABB":{
"pow":%3(sml[1]+offset)%,
"cpow":%2dval%
}

; hide all meter values from web ui
>M 2
+1,4,c,1,-50,ABB
+2,22,m,0,2400,Modbus,21,2,0104000C,0104004A
1,1-0:1.8.0*255(@1000,*,kWh,Energy_Cnt,4
1,=d 1 10 @1,B21 Current Power,W,Power_curr,2
2,010404ffffffff@i1:1,*,kWh,SDM120-1 Energy_out,3
2,010404ffffffff@i0:-1,*,W,SDM120-1 Power_act,2
#

![Cal6](https://user-images.githubusercontent.com/33256321/178226769-f072f5fd-babb-406f-81be-3820aa210bfe.JPG)

**Comment by gemu2015** [comment, marked as answer]
accuracy is dependent on delta time, the longer the better

you may also try to use time between pulses, insert this extra line to test

1,1-0:1.7.0\*255(@1,time,ms,Energy_time,0

this measures time between pulses in ms

you will have to convert this to W , don't know what the factor might be

**Reply by h-tro** [reply]
1.1-0:1.7.0*255(@1,time,ms,Energy_time,0
indicates the number of pulses per minute.
The value increases when more pulses are sent.
60 W = 1 (pulse) 120 W= 2 (pulses).

It's not the time between pulses

**Reply by gemu2015** [reply]
sorry is frequency

**Reply by h-tro** [reply]
Hi gemu,

1,1-0:1.7.0*255(@1,B21 pulses/minute,,pulses/minute,4 calculates pulses per minute based on the time between two pulses.
This is very accurate and reliable (see screenshot). From this you can easily calculate the current power.
The problem is when the power goes to zero, the next pulse is missed for a calculation and the last calculated value is displayed permanently. 
**Please, can you add a timer to the script that sets "pval" to zero after e.g. 5 minutes if there are no pulses???**

Script:
>D
; additional defines
;#define USE_SML_SCRIPT_CMD
;#define USE_SCRIPT_WEB_DISPLAY
;#define USE_SCRIPT_JSON_EXPORT

; offset=16.688
offset=0
lval=0
dval=0
pval=0

>B
->sensor53 r

>S
; calc 60 second delta
if upsecs%60==0{
dval=sml[1]-lval*60*1000
lval=sml[1]
}

if upsecs%10==0{
pval=sml[3]*6
}

; show values in web ui
>W
B21 Energy{m}%4(sml[1]+offset)% kWh
B21 Current Power_dval{m}%4dval% W
B21 Current Power_pval{m}%4pval% W
SDM120-1 Energy{m}%3sml[4]% kWh
SDM120-1 Curr Power{m}%2sml[5]% W

; send calculated values to MQTT
>J
,"ABB":{
"pow":%3(sml[1]+offset)%,
"cpow":%2dval%
}

; hide all meter values from web ui
>M 2
+1,4,c,1,-50,ABB
+2,22,m,0,2400,Modbus,21,2,0104000C,0104004A
1,1-0:1.8.0*255(@10000,*,kWh,Energy_Cnt,4
1,=d 1 10 @1,B21 Current Power,W,Power_curr,2
1,1-0:1.7.0*255(@1,B21 pulses/minute,,pulses/minute,4
2,010404ffffffff@i1:1,*,kWh,SDM120-1 Energy_out,3
2,010404ffffffff@i0:-1,*,W,SDM120-1 Power_act,2
#

![Cal8](https://user-images.githubusercontent.com/33256321/178956240-92bb6fe4-da5e-4db0-b481-852def8edad8.JPG)

**Reply by gemu2015** [reply]
try this. if channel is valid it is updated, else set to zero  
(smlv[channel] is >0 when channel was updated, reading smlv[0] resets all valid flags)

if upsecs%10==0 {
	if smlv[3]>0 {
		; this resets all valid flags
		pval=smlv[0]
		pval=sml[3]*6
	} else {
		pval=0
	}
}
